### ISSUE TITLE
S3Queue silently loses consumed rows on power loss (at-least-once becomes at-most-once; unrecoverable with after_processing='delete')

### ISSUE THREAD

**Issue by zlareb1** [ClickHouse/ClickHouse#111537]
### Describe what's wrong

`S3Queue` (`StorageObjectStorageQueue`) can **silently and permanently lose already-consumed rows** after a power loss of the ClickHouse node, converting its documented at-least-once delivery into at-most-once. With `after_processing = 'delete'` the loss is **unrecoverable** because the source object is deleted before the ingested rows are durable.

The root cause is a cross-domain ordering gap in `StorageObjectStorageQueue::commit`:

1. `streamToViews` runs the insert pipeline, so the consumed rows land in the target `MergeTree` part — in the **page cache only** (no fsync; `fsync_after_insert` is 0 by default and there is no `S3Queue` setting to order durability before the commit). `src/Storages/ObjectStorageQueue/StorageObjectStorageQueue.cpp:934-935,980`
2. Inside `commit`, when `after_processing != KEEP`, the source objects are **deleted** (`removeObjectsIfExist`). `StorageObjectStorageQueue.cpp:1049-1053` → `src/Storages/ObjectStorageQueue/ObjectStorageQueuePostProcessor.cpp:75`
3. Then the "processed" nodes are committed to **Keeper** via `tryMulti`. `StorageObjectStorageQueue.cpp:1078-1079`

Keeper's commit is force-synced and, being a separate service (or a separate node), is unaffected by *this* node's power loss. So if the ClickHouse node loses power after step 3 but before the target part is fsynced, on restart Keeper reports the files as already processed, `S3Queue` never re-reads them, and the rows are gone. This is a device-level power loss (dirty page-cache loss); a plain `SIGKILL` does **not** expose it because the page cache survives a process kill — which is why normal crash testing misses it.

This contradicts the documented behavior: the "Limitations" section of the `S3Queue` docs lists "abnormal server termination" **only** as a cause of *duplicated* rows, never loss (`docs/en/engines/table-engines/integrations/s3queue.md`, "Limitations").

The same ordering exists in the `Kafka` engine (broker offset committed after an un-fsynced insert) and `FileLog` (un-fsynced `.meta` offset marker), but `S3Queue` is the clearest case because Keeper is a genuinely independent durability domain and `after_processing = 'delete'` destroys the only other copy.

### Does it reproduce on the most recent release?

Yes — reproduced on `26.4.1.1` and on a recent `master` build `26.7.1.1258` (identical result). The code path is present on current `master`.

### How to reproduce

Topology: only the ClickHouse node loses power; Keeper and the object store stay alive (the realistic case — Keeper/S3 are separate services).

- ClickHouse server `<path>` on a filesystem backed by a `dm-flakey` device (faithful power loss = drop all un-fsynced writes).
- Keeper as a separate process on a normal filesystem (survives).
- MinIO on a normal filesystem (survives).

Steps:

1. `CREATE TABLE tgt (id UInt64, val UInt64) ENGINE = MergeTree ORDER BY id;` (default settings, `fsync_after_insert = 0`)
2. `CREATE TABLE q (id UInt64, val UInt64) ENGINE = S3Queue('http://minio/bkt/*.csv', 'user', 'pass', 'CSV') SETTINGS mode = 'unordered', keeper_path = '/s3q', after_processing = 'delete';`
3. `CREATE MATERIALIZED VIEW mv TO tgt AS SELECT id, val FROM q;`
4. Upload N files to the bucket. Wait until all rows are ingested into `tgt` and the processed nodes appear under `keeper_path/processed`.
5. Cut power to the ClickHouse node (drop all un-fsynced writes) — Keeper and MinIO stay up.
6. Restart the ClickHouse node.

Result: `SELECT count() FROM tgt` is 0 (all consumed rows lost); the processed nodes still exist in Keeper so `S3Queue` never re-reads the files; with `after_processing = 'delete'` the source objects are already gone → unrecoverable.

A self-contained harness (MinIO + external Keeper + `dm-flakey`, with a fsync-off "canary" table that proves the power-loss window is live, and a matched `fsync_after_insert = 1` control) is available and reproduces this deterministically:

<details><summary>Measured results (20 files × 50 rows = 1000 consumed rows)</summary>

| scenario | target `fsync_after_insert` | outcome |
|---|---|---|
| graceful stop/restart | 0 | 1000/1000 survive (control passes) |
| **power loss (default)** | **0** | **tgt 1000 → 0; Keeper keeps all 20 processed markers; NOT re-ingested; canary lost → window live** |
| power loss | 1 | 1000/1000 survive (part fsynced during `execute()` before the Keeper commit) |

Identical shape with `after_processing = 'delete'`, except the loss is unrecoverable (source deleted at `commit` before the cut).
</details>

### Expected behavior

An acknowledged/consumed file's rows should be durable before `S3Queue` records the file as processed (and, for `after_processing = 'delete'`, before the source object is removed) — i.e. the engine should honor at-least-once across a power loss, or expose a setting to order local durability ahead of the processed-commit. At minimum the loss window should be documented alongside the existing duplicate-only limitation.

### Error message and/or stacktrace

No error — the loss is silent. After recovery `SELECT count() FROM tgt` is simply short, and `system.zookeeper` under `keeper_path/processed` still lists the lost files.

### Additional context

Setting `fsync_after_insert = 1` (and `fsync_part_directory = 1`) on the target table closes the direct-to-table window, but that is off by default, is not documented as required for `S3Queue` correctness, and does not help an operator who relies on the documented at-least-once guarantee. Found by device-level power-loss testing (dm-flakey); invisible to `SIGKILL`-based crash tests because the page cache survives a process kill.

### Version info
- Resolved by: #111571
- Merged into: `26.8.1.90` (included in `26.8` and later)

**Comment by zlareb1** [comment]
@groeneai fix

**Comment by groeneai** [comment]
Acknowledged. Investigating the S3Queue commit-ordering durability gap in `StorageObjectStorageQueue::commit`: the target MergeTree part is durable in the page cache only, while source objects are deleted and the processed nodes are committed to Keeper before that part is fsynced, so a device-level power loss between the Keeper commit and the fsync drops already-consumed rows. I'll follow up here with a PR or my findings.

**Comment by groeneai** [comment]
Opened #111571 for the documentation part of this: it documents the power-loss loss window in the `S3Queue`/`AzureQueue` "Limitations" section (which previously listed abnormal termination only as a cause of *duplicated* rows) and the `fsync_after_insert = 1` + `fsync_part_directory = 1` mitigation on the target table, scoped to the materialized-view consumption path where `commit` runs only after the whole insert pipeline finishes.

On the code side: `commit` cannot force target durability itself, because `fsync_after_insert` is a `MergeTree` table setting and the engine inserts into arbitrary dependent-MV targets, and the source-delete-before-Keeper-commit ordering is deliberate (multi-replica consistency, `ObjectStorageQueuePostProcessor.cpp` / `ObjectStorageQueueSource.cpp`). A real fix would be a new opt-in engine setting that fsyncs the touched target parts before the processed-commit (and, for `after_processing='delete'`, before the source removal), which is a cross-engine durability-contract decision (`Kafka` and `FileLog` share the same offset-after-un-fsynced-insert ordering) with backward-compatibility implications. I've left that design to the component owner rather than deciding the semantics unilaterally in this PR. Happy to implement it if there's a preferred shape.

**Comment by zlareb1** [comment]
@groeneai what about the fix PR? We don't need it now?

**Comment by groeneai** [comment]
Status: the docs PR #111571 is merged and kssenii closed this as completed, so #111537 is resolved on the docs side (the S3Queue/AzureQueue Limitations sections now cover the power-loss window and the `fsync_after_insert = 1` target-table mitigation).

There is no code fix yet. A code fix would be a new opt-in engine setting that fsyncs the touched target parts before the processed-commit (and, for `after_processing = 'delete'`, before the source removal). That is a cross-engine durability-contract decision with backward-compatibility implications, and the Kafka (#111799) and RabbitMQ (#111800) issues you filed are the same family, so I left the semantics to the component owner rather than deciding them unilaterally. kssenii resolved this one with the docs.

If the team wants that setting, I am happy to implement whatever shape kssenii prefers; the cross-engine issues are the natural place to settle the design.

**Comment by zlareb1** [comment]
lets implement it @groeneai

**Comment by groeneai** [comment]
Acknowledged, on it. I'll implement the opt-in setting (fsync the touched target parts before the processed-commit, and before source removal for after_processing='delete', default off for backward-compat) and open a PR here for kssenii to review; I'll follow up in this thread with the PR link, or my findings if the engine-side fsync mechanism turns out to need a design call first.

**Comment by groeneai** [comment]
Opened #111861 with the first, mechanism-independent piece of this fix.

It closes a durability gap in `DataPartStorageOnDiskBase::rename`: with `fsync_part_directory = 1` only the moved part directory was fsynced, not the parent directory that holds the renamed dentry, so the `tmp_* -> final` rename was not itself crash-durable. A power cut just after a part was committed could leave the part missing on disk even though its files were already written. The PR fsyncs the destination parent (and, for cross-parent moves, the source parent) under the same setting, so `fsync_after_insert = 1` + `fsync_part_directory = 1` on the target now actually makes the consumed part durable, including its rename.

This does not by itself change the engine-side ordering (when S3Queue/Kafka/RabbitMQ commit the source offset or delete the source object relative to that durable insert). That cross-engine "durable target insert before source ack" contract is the larger change, shared with #111799 (Kafka) and #111800 (RabbitMQ); I'm tracking the mechanism decision there and will follow up once the shape is settled.
